feature/add hpmicro device support

willChuai opened the pull request:
This PR adds support for HPMicro HPM6750 SoC and hpm6750evkmini board that is use the SoC.

- Add HPM6750 SoC support.
- Add HPM6360 SoC support.
- Add HPM6750EVKMINI support
- Drivers: Serial, GPIO,Pinctrl

hal_hpmicro module request:
#46301

fkokosinski commented on boards/riscv/hpm6750evkmini/doc/index.rst:
```diff
@@ -0,0 +1,109 @@
+It has a 2MB continuous on-chip ram.
+Also, it provides various memory interfaces, including SDRAM, Quad SPI NOR Flash, SD/eMMC.
+
+The figure shows the PCB is HPM6750EVKmini platform.
```
`The figure shows the PCB of the HPM6750EVKmini platform`?

fkokosinski commented on boards/riscv/hpm6750evkmini/doc/index.rst:
```diff
@@ -0,0 +1,109 @@
+Hardware
+********
+
+The HPM6750EVKMINI platform integrates 2 cores 32-bit 816MHz RISC-V CPUs, DSP,
```
Nitpick: `HPM6750EVKMINI` was previously spelled as `HPM6750EVKmini`. This documentation could use some consistency here.

fkokosinski commented on boards/riscv/hpm6750evkmini/hpm6750evkmini-pinctrl.dtsi:
```diff
@@ -0,0 +1,20 @@
+ * Copyright 2022 hpmicro
+ * SPDX-License-Identifier: Apache-2.0
+ */
+ #include <zephyr/dt-bindings/pinctrl/hpmicro-pinctrl-common.h>
```
Unnecessary white space before this line

fkokosinski commented on boards/riscv/hpm6750evkmini/hpm6750evkmini.yaml:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+identifier: hpm6750evkmini
+name: HPMicro hpm6750evkmini
```
Make this name consistent with the documentation

fkokosinski commented on boards/riscv/hpm6750evkmini/hpm6750evkmini_defconfig:
```diff
@@ -0,0 +1,25 @@
+CONFIG_GPIO=y
+
+# HW DSP options
+#CONFIG_SOC_ANDES_V5_HWDSP=y
```
Commented-out config option?

willChuai replied:
Using dsp requires the user to provide a toolchain that supports the dsp command.

fkokosinski replied:
Sure. Why isn't `CONFIG_SOC_ANDES_V5_HWDSP` defined in any Kconfig file? Or am I missing something?

willChuai replied:
There was a problem with the first commit, and the second commit fixed the problem with the file `soc\riscv\riscv-privilege\hpmicro\Kconfig.soc`

fkokosinski commented on boards/riscv/hpm6750evkmini/hpm6750evkmini_defconfig:
```diff
@@ -0,0 +1,25 @@
+#CONFIG_SOC_ANDES_V5_HWDSP=y
+
+# Cache options
+# CONFIG_CACHE_ENABLE=y
```
Ditto

willChuai replied:
I will remove it, flash mode is not currently supported and cache is not working for now. flash support will be committed after the initial release and cache will be turned on at that time.

fkokosinski commented on boards/riscv/hpm6750evkmini/hpm6750evkmini_defconfig:
```diff
@@ -0,0 +1,25 @@
+# CONFIG_CACHE_ENABLE=y
+
+# Nocache memory options
+#CONFIG_NOCACHE_MEMORY=y
```
Ditto.

willChuai replied:
Again, cache is not currently enabled and I will remove this code

fkokosinski commented on drivers/gpio/Kconfig.hpmicro:
```diff
@@ -0,0 +1,12 @@
+config GPIO_HPMICRO
+	bool "HPMicro GPIO driver"
+	depends on (SOC_SERIES_HPMICRO)
+	depends on (HAS_HPMSDK_GPIO)
```
Why are `SOC_SERIES_HPMICRO` and `HAS_HPMSDK_GPIO` in parentheses?

fkokosinski commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,261 @@
+ *
+ */
+
+
```
Unnecessary blank line.

fkokosinski commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,261 @@
+#include "gpio_utils.h"
+
+struct gpio_hpm_config {
+	/* gpio_driver_config needs to be first */
```
Why?

fkokosinski commented on drivers/serial/Kconfig.hpmicro:
```diff
@@ -0,0 +1,15 @@
+# Copyright 2022 hpmicro
+# SPDX-License-Identifier: Apache-2.0
+
+
```
Unnecessary blank line.

fkokosinski commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,322 @@
+void uart_hpm_irq_err_disable(const struct device *dev)
+{
+
+}
```
A small comment would be nice here to explain why these functions are empty. I assume it is because this UART driver doesn't support these features.

fkokosinski replied:
I don't think this had been addressed?

willChuai replied:
Sorry, soc doesn't provide this function, should I keep this function or add a comment. Currently my approach is to add comments.

fkokosinski commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,322 @@
+	const struct uart_hpm_cfg *const cfg = dev->config;
+
+	if ((uart_get_irq_id((UART_Type *)cfg->base) & uart_intr_id_rx_data_avail) ||
+	(uart_get_irq_id((UART_Type *)cfg->base) & uart_intr_id_tx_slot_avail)) {
```
Don't call `uart_get_irq_id` two times here, please. Use a variable instead.

fkokosinski commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,322 @@
+	}
+
+#else
+/* !CONFIG_UART_INTERRUPT_DRIVEN */
```
Merge this comment with the preceding line.

fkokosinski reviewed (commented):
More general questions:

1. Some files use copyright notice with 2021, some with 2022. Is this intentional?
2. You're adding a lot of SoCs, but only one board that uses only one of them. Why is that?

willChuai commented:
> More general questions:
> 
> 1. Some files use copyright notice with 2021, some with 2022. Is this intentional?
> 2. You're adding a lot of SoCs, but only one board that uses only one of them. Why is that?

1. Yes, the file with the copyright notice 2021 is a board-level support file, copied directly from hpm_sdk and can be modified if necessary.
2.  Another soc also has a supported  board and is planned to be integrated into zephyr.

fkokosinski commented on drivers/gpio/Kconfig.hpmicro:
```diff
@@ -0,0 +1,12 @@
+	depends on SOC_SERIES_HPMICRO
+	depends on HAS_HPMSDK_GPIO
+	default $(dt_compat_enabled,$(DT_COMPAT_HPMICRO_HPM_GPIO))
+	select HAS_DTS_GPIO
```
Checkpatch complains that this config option doesn't exist.

gmarull replied:
this option was recently removed

gmarull commented on modules/hal_hpmicro/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+set(HPM_BOARD_DIR ${CONFIG_BOARD})
+string(TOUPPER ${SOC_NAME} HPM_SOC_TOUPPER)
+set(HPM_SOC ${HPM_SOC_TOUPPER})
+set(ENV{GNURISCV_TOOLCHAIN_PATH} ${SYSROOT_DIR})
```
cc: @tejlmand

gmarull commented on modules/hal_hpmicro/Kconfig:
```diff
@@ -0,0 +1,80 @@
+# SPDX-License-Identifier: Apache-2.0
+config ZEPHYR_HAL_HPMICRO_MODULE
+	bool
+config HAS_HPMICRO_HAL
+	bool "HPMSDK HAL drivers support"
+config HAS_HPMSDK_UART
+	bool "HPMSDK UART driver support"
+config HAS_HPMSDK_DRAM
+	bool "HPMSDK DRAM driver support"
+config HAS_HPMSDK_SDP
+	bool "HPMSDK SDP driver support"
```
options to enable certain HAL features are tipically named "USE_", to indicate e.g. that a driver uses certain HAL modules. "HAS_" is tipically used to indicate a certain SoC has a feature. You can check GD32/STM32 for reference.

willChuai replied:
Yes, these configurations are passed to hpm_sdk for the purpose of trimming the functionality of the soc, so using `HAS` here I think would make more sense, again https://github.com/zephyrproject-rtos/zephyr/pull/46329#discussion_ r895501958 's comment is the same as this one, HAS_X is placed at the soc to indicate whether the SOC supports this feature.

gmarull commented on modules/hal_hpmicro/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+
+if(CONFIG_HAS_HPMICRO_HAL)
+
+set(HPM_SDK_BASE ${ZEPHYR_HAL_HPMICRO_MODULE_DIR}/hpm_sdk)
+set(HPM_BOARD_DIR ${CONFIG_BOARD})
+string(TOUPPER ${SOC_NAME} HPM_SOC_TOUPPER)
+set(HPM_SOC ${HPM_SOC_TOUPPER})
```
please indicate where are these used.

willChuai replied:
A module named `hal_hpmicro` has been added, which internally relies on these macros to control the behaviour of cmake. This format is good for code maintenance, if `hpm_sdk` is updated it can be copied over directly without having to change anything.

gmarull replied:
why isn't this done at HAL level then? or all in `modules/`? It feels like CMake is splitted in 2, making it hard to know what's going on here.

willChuai replied:
The cmake for hal_hpmicro is the cmake for hpm_sdk, which requires some parameters to be passed in by the user when using it. After being integrated into zephyr, we want to be able to maintain compatibility with hpm_sdk without having to maintain an additional environment.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2022 hpmicro
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_include_directories(${CONFIG_SOC})
```
what does this include, and why?

gmarull commented on soc/riscv/riscv-privilege/hpmicro/Kconfig.soc:
```diff
@@ -0,0 +1,62 @@
+	select RISCV_ISA_EXT_M
+	select RISCV_ISA_EXT_A
+	select RISCV_ISA_RV32I
+	select HAS_HPMSDK_UART
+	select HAS_HPMSDK_GPIO
+	select HAS_HPMSDK_PINCTRL
+	select HAS_HPMSDK_PDM
+	select HAS_HPMSDK_PLLCTL
+	select HAS_HPMSDK_PMP
```
drivers need to select what they need, not SoC.

gmarull replied:
ping, these options should not be here but in drivers.

willChuai replied:
Okay, I'll remove the driver configuration, but I need to keep the soc configuration.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/Kconfig.soc:
```diff
@@ -0,0 +1,62 @@
+prompt "FPU options"
+default SOC_HPMICRO_NO_FPU
+
+config SOC_HPMICRO_NO_FPU
+	bool "No FPU"
+
+config SOC_HPMICRO_SINGLE_PRECISION_FPU
+	bool "Single precision FPU"
+	select CPU_HAS_FPU
+
+config SOC_HPMICRO_DOUBLE_PRECISION_FPU
+	bool "Double precision FPU"
+	select CPU_HAS_FPU_DOUBLE_PRECISION
```
why not use options like `CPU_HAS_FPU_DOUBLE_PRECISION` at SoC level, see e.g. how STM32H7 does this selection.

gmarull replied:
ping

willChuai replied:
We want to tell users that the chip supports both single and double precision fpu.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.c:
```diff
@@ -0,0 +1,49 @@
+ */
+
+#include <init.h>
+extern void board_init_ahb(void);
```
why do we need board specific hooks? can you please explain?

gmarull replied:
ping, no response here.

willChuai replied:
Initialized the soc clock and placed it under the board in hpm_sdk. I wanted to reuse the code so I called it directly.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef __RISCV_HPMICRO_SOC_H_
+#define __RISCV_HPMICRO_SOC_H_
+
+#include <soc_common.h>
+#include <zephyr/devicetree.h>
+
+/* Machine timer memory-mapped registers */
+#define RISCV_MTIME_BASE             DT_REG_ADDR_BY_IDX(DT_NODELABEL(mtimer), 0)
+#define RISCV_MTIMECMP_BASE          DT_REG_ADDR_BY_IDX(DT_NODELABEL(mtimer), 1)
+
+/* lib-c hooks required RAM defined variables */
+#define RISCV_RAM_BASE               CONFIG_SRAM_BASE_ADDRESS
+#define RISCV_RAM_SIZE               KB(CONFIG_SRAM_SIZE)
+
+/* Include CSRs available for Andes V5 SoCs */
+#ifndef _ASMLANGUAGE
+#include <toolchain.h>
+#include <hpm_common.h>
+#include <hpm_soc.h>
+#include <board.h>
+#endif  /* !_ASMLANGUAGE */
```
please don't use `soc.h` as a _general purpose_ header. Some options may need to be kept until other modules are refactored, but I guess most of this can be removed and defined/included elsewhere.

ref. https://github.com/zephyrproject-rtos/zephyr/pull/46147

willChuai replied:
Okay, remove unnecessary files.

gmarull commented on dts/riscv/hpmicro/hpm6750.dtsi:
```diff
@@ -0,0 +1,246 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+		timebase-frequency = <816000000>;
+		CPU0: cpu@0 {
```
no `UPPERCASE` node labels please

gmarull commented on dts/riscv/hpmicro/hpm6750.dtsi:
```diff
@@ -0,0 +1,246 @@
+ */
+
+/dts-v1/;
+#include <dt-bindings/gpio/gpio.h>
```
please prefix all zephyr includes with `zephyr/`

gmarull commented on dts/riscv/hpmicro/hpm6750.dtsi:
```diff
@@ -0,0 +1,246 @@
+	};
+};
+
+&gpio0 {
+	gpioa: gpio@0 {
+		compatible = "hpmicro,hpm-gpio";
+		reg = <0x0 0x4000>;
+		gpio-controller;
+		#gpio-cells = <2>;
+		hpmicro-gpio-port = <0>;
+		interrupts = <1 1>;
+		interrupt-parent = <&plic0>;
+		label = "GPIOA";
+	};
```
why not define these entries after `gpio0` definition above?

willChuai replied:
On this chip gpioa-gpioz can be assigned to any of the gpio controllers gpio0, gpio1 and fpgio.

gmarull commented on drivers/pinctrl/Kconfig.hpmicro:
```diff
@@ -0,0 +1,12 @@
+
+config PINCTRL_HPMICRO
+	bool "Pin controller driver for HPMicro MCUs"
+	depends on (SOC_SERIES_HPMICRO)
+	depends on (HAS_HPMSDK_PINCTRL)
```
parentheses are redundant here. note that the HAL options are tipically named "USE_" as mentioned earlier.

gmarull commented on drivers/pinctrl/pinctrl_hpmicro.c:
```diff
@@ -0,0 +1,64 @@
+#include <soc.h>
+#include <zephyr/drivers/pinctrl/pinctrl_hpmicro_common.h>
+
+#include "hpm_common.h"
+#include "hpm_soc.h"
```
please use `<>` for non-local headers.

gmarull commented on include/zephyr/drivers/pinctrl/pinctrl_hpmicro_common.h:
```diff
@@ -0,0 +1,47 @@
+/*
```
any reason to not use `pinctrl_soc.h` directly? This is generally done when multiple SoCs share pinctrl drivers, like for example GD32 ARM/RISC-V SoCs.

willChuai replied:
Yes, we plan to add another series of zephyr's chip pinctrl modules that will have some differences, this way we can ensure that we can reuse the driver.

gmarull commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,55 @@
+# Copyright 2022 hpmicro
+# SPDX-License-Identifier: Apache-2.0
+
+description: HPMicro PinMux
```
please extend this documentation with some more details, check other vendors as a reference.

gmarull commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,55 @@
+      power-source:
+        required: false
+        type: string
+        default: "v3_3"
+        enum:
+          - "v3_3"
+          - "v1_8"
```
suggestion: `3v3` and `1v8`

gmarull commented on include/zephyr/drivers/pinctrl/pinctrl_hpmicro_common.h:
```diff
@@ -0,0 +1,47 @@
+#ifndef ZEPHYR_INCLUDE_DRIVERS_PINCTRL_PINCTRL_HPMICRO_COMMON_H_
+#define ZEPHYR_INCLUDE_DRIVERS_PINCTRL_PINCTRL_HPMICRO_COMMON_H_
+
+#define HPMICRO_PAD_NUM(_mux)\
+(((_mux) >> HPMICRO_PIN_NUM_SHIFT) & HPMICRO_PIN_NUM_MASK)
```
nit: improve formatting here,
```c
#define HPMICRO_PAD_NUM(_mux) \
	(((_mux) >> HPMICRO_PIN_NUM_SHIFT) & HPMICRO_PIN_NUM_MASK)
```

gmarull commented on include/zephyr/dt-bindings/pinctrl/hpmicro-pinctrl-common.h:
```diff
@@ -0,0 +1,51 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_HPMICRO_PINCTRL_COMMON_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_HPMICRO_PINCTRL_COMMON_H_
+
+#include <zephyr/sys/util_macro.h>
```
this is not required by this file.

gmarull commented on include/zephyr/dt-bindings/pinctrl/hpmicro-pinctrl-common.h:
```diff
@@ -0,0 +1,51 @@
+
+#include <zephyr/sys/util_macro.h>
+
+#define HPMICRO_PIN_NUM_SHIFT      0U
+#define HPMICRO_PIN_NUM_MASK       0x3FFU
+
+
+#define HPMICRO_PIN_ANALOG_MASK      1U
+#define HPMICRO_PIN_ANALOG_SHIFT     10U
+#define HPMICRO_PIN_ALT_MASK         0x1FU
+#define HPMICRO_PIN_ALT_SHIFT        11U
+
+#define HPMICRO_PINMUX(pin, is_analog, alt)						\
+		(((pin & HPMICRO_PIN_NUM_MASK) << HPMICRO_PIN_NUM_SHIFT) |		\
+		((is_analog & HPMICRO_PIN_ANALOG_MASK) << HPMICRO_PIN_ANALOG_SHIFT) |	\
+		((alt & HPMICRO_PIN_ALT_MASK) << HPMICRO_PIN_ALT_SHIFT))
+
+#define HPMICRO_OPEN_DRAIN			1U
+#define HPMICRO_OPEN_DRAIN_SHIFT	0U
+
+#define HPMICRO_NO_PULL				1U
+#define HPMICRO_NO_PULL_SHIFT		1U
+
+#define HPMICRO_PULL_DOWN			1U
+#define HPMICRO_PULL_DOWN_SHIFT		2U
+
+#define HPMICRO_PULL_UP				1U
+#define HPMICRO_PULL_UP_SHIFT		3U
+
+#define HPMICRO_FORCE_INPUT				1U
+#define HPMICRO_FORCE_INPUT_SHIFT		4U
+
+#define HPMICRO_DRIVER_STRENGTH				7U
+#define HPMICRO_DRIVER_STRENGTH_SHIFT		5U
+
+#define HPMICRO_SCHMITT_ENABLE				1U
+#define HPMICRO_SCHMITT_ENABLE_SHIFT		12U
+
+#define HPMICRO_POWER			1U
+#define HPMICRO_POWER_SHIFT		13U
```
can you please document these macros? specially those meant for public use.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/pinctrl_soc.h:
```diff
@@ -0,0 +1,63 @@
+#ifndef ZEPHYR_SOC_RISCV_RISCV_PRIVILEGE_HPMICRO_PINCTRL_SOC_H_
+#define ZEPHYR_SOC_RISCV_RISCV_PRIVILEGE_HPMICRO_PINCTRL_SOC_H_
+
+#include <devicetree.h>
```
ditto, `zephyr/`

gmarull commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,322 @@
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t irq_config_func;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pincfg;
+#endif /* CONFIG_PINCTRL */
```
it doesn't seem your platform has an alternative to pinctrl, so such ifdef should be removed.

gmarull commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,259 @@
+	struct gpio_driver_config common;
+	GPIO_Type *gpio_base;
+	uint32_t  port_base;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pincfg;
+#endif /* CONFIG_PINCTRL */
```
why does a GPIO driver require pinctrl? can you please explain

willChuai replied:
This chip io function is complex, where the pinmux peripheral controls the multiplexing of io functions, drive capability, speed, pull-up and pull-down, analogue or digital functions, etc. The other part is the gpio controller, who is only responsible for input and output and status reading as well as the interrupt controller.

gmarull commented on boards/riscv/hpm6750evkmini/hpm6750evkmini-pinctrl.dtsi:
```diff
@@ -0,0 +1,20 @@
+	};
+	pinmux_gpiob: pinmux_gpiob {
+		group0 {
+			pinmux = <HPMICRO_PINMUX(50, 0, 0)>,
+			<HPMICRO_PINMUX(51, 0, 0)>,
+			<HPMICRO_PINMUX(52, 0, 0)>;
```
nit: please align for readability

gmarull commented on boards/riscv/hpm6750evkmini/hpm6750evkmini.dts:
```diff
@@ -0,0 +1,78 @@
+		zephyr,dtcm = &dlm;
+	};
+
+	core0_lm_slv: memory@1000000 {
+		device_type = "memory";
+		reg = <0x1000000 DT_SIZE_K(512)>;
+	};
+
+	core1_lm_slv: memory@1180000 {
+		device_type = "memory";
+		reg = <0x1180000 DT_SIZE_K(256)>;
+	};
+
+	noncache: memory@11c0000 {
+		device_type = "memory";
+		reg = <0x11c0000 DT_SIZE_K(256)>;
+	};
+
+	dram: memory@40000000 {
+		device_type = "memory";
+		reg = <0x40000000 DT_SIZE_M(16)>;
+	};
```
does this really belong to board level DTS?

willChuai replied:
All changes required except dram

gmarull commented on boards/riscv/hpm6750evkmini/hpm6750evkmini.dts:
```diff
@@ -0,0 +1,78 @@
+	current-speed = <115200>;
+	pinctrl-0 = <&pinmux_uart0>;
+	pinctrl-names = "default";
+	clock-name = <19005459>;
+	clock-src = <0>;
```
?

willChuai replied:
The clock name is an enumerated type defined in the hpm driver and each soc has its own header file to define it, currently it can be linked to the appropriate file correctly but I don't currently have a way to call the header file in the hal modules directly from within the dts.

galak replied:
So you should probably implement a clock controller driver in zephyr.  Its unfortunate that you guys use "clock name" to mean a clock identifier.  Its going to be very confusing to use a property called `clock-name` as that has a well defined meaning in Zephyr already.

As this should just be a `clocks` property you can do something like:

`clocks = <&clk_ctrl sysctl_resource_uart0 CLK_SRC_GROUP_COMMON clock_node_uart0 0>`;

where the cells are:
- resource
- src_type
- node
- src

or:

`clocks = <&clk_ctrl clock_uart0 0>;

Where the cells are:
- name
- src

gmarull reviewed (changes requested):
thanks for your contribution! please check my comments after an initial review

willChuai commented:
> thanks for your contribution! please check my comments after an initial review

Thanks, I will revise it as soon as possible, it will take some time.

keith-zephyr commented on boards/riscv/hpm6750evkmini/doc/index.rst:
```diff
@@ -0,0 +1,109 @@
+the :ref:`hello_world` application.
+
+.. zephyr-app-commands::
+   :board: hpm6750evkmini
```
Does this also need `:zephyr-app: samples/hello_world`?

willChuai replied:
yes

keith-zephyr commented on boards/riscv/hpm6750evkmini/doc/index.rst:
```diff
@@ -0,0 +1,109 @@
+Flashing
+========
+
+ ``CONFIG_XIP=n``, you can load the program (``zephyr.elf``) into RAM directly and execute it.
```
`CONFIG_XIP=n` is the default setting correct?  Can you clearly state that here?  As is, this implies the user should override this setting before building.

willChuai replied:
This version does not yet enable xpi flash, so it can only run code in ram at the moment.

keith-zephyr replied:
If XIP flash is not supported, please state that in the documentation.
For example:
```
Execute-in place flash is not support (CONFIG_XIP-=n), so the program (zephyr.elf) is loaded into and executed directly from RAM.
```

willChuai commented on boards/riscv/hpm6750evkmini/doc/index.rst:
```diff
@@ -0,0 +1,109 @@
+********
+
+The HPM6750EVKMINI platform integrates 2 cores 32-bit 816MHz RISC-V CPUs, DSP,
+2M RAM, Cache, SPI flash memory, ethernet controller and other peripherals.
```
Yes, it is

willChuai replied:
2Mbyte

willChuai commented on boards/riscv/hpm6750evkmini/doc/index.rst:
```diff
@@ -0,0 +1,109 @@
+
+- HPM6750IVM MCU (816Mhz, 2MB OCRAM)
+- Onboard Memory
+  - 128Mb SDRAM
```
128Mbit

willChuai commented on boards/riscv/hpm6750evkmini/doc/index.rst:
```diff
@@ -0,0 +1,109 @@
+Hardware
+********
+
+The HPM6750EVKMINI platform integrates 2 cores 32-bit 816MHz RISC-V CPUs, DSP,
```
No problem, we'll add the link

gmarull commented on modules/hal_hpmicro/Kconfig:
```diff
@@ -0,0 +1,86 @@
+config HAS_HPMICRO_HAL
+	bool "HPMSDK HAL drivers support"
+
+if HAS_HPMICRO_HAL
+
+config HAS_HPMSDK_UART
+	bool "HPMSDK UART driver support"
+config HAS_HPMSDK_DRAM
+	bool "HPMSDK DRAM driver support"
+config HAS_HPMSDK_SDP
```
again, `HAS_` doesn't look right. These options are being used to device which HAL drivers are compiled here https://github.com/hpmicro/zephyr_hal_hpmicro/blob/manifest-rev/hpm_sdk/drivers/CMakeLists.txt, please, to be consistent with other platforms use `USE_`. Drivers should be responsible to select what they need. For example, the UART driver will `select USE_HPMSDK_UART`.

willChuai replied:
OK, will make changes.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/linker.ld:
```diff
@@ -0,0 +1,33 @@
+ *
+ */
+#include <linker/linker-tool.h>
+#include <devicetree.h>
```
missing prefix

keith-zephyr commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2022 hpmicro
```
Can this driver (and the pinctrl and uart drivers) get added directly into the main tree?

keith-zephyr replied:
Whoops - ignore this comment.  I see this commit is now adding to the main tree instead of a separate HAL.

galak commented on drivers/pinctrl/pinctrl_hpmicro.c:
```diff
@@ -0,0 +1,64 @@
+
+#include <hpm_common.h>
+#include <hpm_soc.h>
+#define IOC_BASE_ADDRESS DT_REG_ADDR(DT_NODELABEL(pinctrl))
```
This should be `DT_INST_REG_ADDR(0)`, not `DT_REG_ADDR(DT_NODELABEL(pinctrl))`

willChuai replied:
> Why was this marked as resolved, without addressing the comment or stating why it shouldn't apply?

There's nothing wrong with writing it that way. Under the current module, other chips are also made in this way, if there is a problem, please change it uniformly.
[esp32](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pinctrl/pinctrl_esp32.c)
[NXP](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pinctrl/pinctrl_kinetis.c)
[GD](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pinctrl/pinctrl_gd32_afio.c)

galak commented on dts/bindings/gpio/hpmicro,hpm-gpio.yaml:
```diff
@@ -0,0 +1,40 @@
+    reg:
+      required: true
+
+    label:
```
remove `label` property.

galak commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,103 @@
+          - pin: The gpio pin number (0, 1, ..., GPIO_NUM_MAX)
+          - analog: bool, select analog pin in pad
+          - alt_select: alt0-alt31
+      drive-strength:
```
This doesn't look right.  I'm assuming these are binary values for the register field or something.  Don't these values mean something?  How would anyone know how to set this field?

galak commented on dts/bindings/serial/hpmicro,hpm-uart.yaml:
```diff
@@ -0,0 +1,25 @@
+    clock-name:
+      type: int
+      description: Uart clock name
+      required: true
```
Why is `clock-name` an int?  Also, why isn't this coming from base.yaml?

galak commented on dts/bindings/serial/hpmicro,hpm-uart.yaml:
```diff
@@ -0,0 +1,25 @@
+      description: Uart clock name
+      required: true
+
+    clock-src:
```
This should just be `clocks` coming from base.yaml.

galak commented on dts/riscv/hpmicro/hpm6750.dtsi:
```diff
@@ -0,0 +1,251 @@
+};
+
+&gpio0 {
+
```
Why is this just not with `gpio0: peripheral@f0000000 {...` above.

galak reviewed (changes requested):
A few devicetree things here are there.

Also, please look at updating the driver/*/Kconfig.hpmicro to use the newer style.  We are replacing `DT_COMPAT_HPM_UART` and `default $(dt_compat_enabled,$(DT_COMPAT_HPM_UART))` pattern.

Now we auto generate the `DT_COMPAT_HPM_UART` variable so you don't need to define that anymore (we do this for all `compatible` properties we find in the YAML bindings).

Instead of `$(dt_compat_enabled,$(DT_COMPAT_HPM_UART))` we are now doing:

```
   default y
   depends on DT_HAS_HPMICRO_HPM_UART_ENABLED
```

(See `drivers/serial/Kconfig.numicro` for an example).  This applies to GPIO, pinctrl, serial, etc.

galak commented on drivers/serial/Kconfig.hpmicro:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2022 hpmicro
+# SPDX-License-Identifier: Apache-2.0
+DT_COMPAT_HPM_UART := hpmicro,hpm-uart
```
Using new style, this should become:

```
# Copyright 2022 hpmicro
# SPDX-License-Identifier: Apache-2.0

config UART_HPM
	bool "HPMICRO serial driver"
	default y
	depends on DT_HAS_HPMICRO_HPM_UART_ENABLED
	select USE_HPMSDK_UART
	select SERIAL_HAS_DRIVER
	select SERIAL_SUPPORT_INTERRUPT
	help
	  This option enables the UART driver for HPM SoC family.
```

(similar change needed for GPIO & Pinctrl Kconfig's)

willChuai commented:
> Thanks for these additions. In the meantime the PR has grown quite significantly, could you please keep the change-set in a reasonable size to ease the review process. I'd suggest to limit the first PR to basic SoC support including serial, gpio, and pinctrl drivers, and add further peripherals in follow-up PRs.

Thanks, great suggestion, I will split the code.

galak commented on dts/bindings/clock/hpmicro,hpm-clock.yaml:
```diff
@@ -0,0 +1,45 @@
+    reg:
+      required: true
+
+    clock-core-name:
```
rename this property to remove `name`, doesn't make sense if the type is int.

galak commented on dts/bindings/gpio/hpmicro,hpm-gpio.yaml:
```diff
@@ -0,0 +1,37 @@
+    reg:
+      required: true
+
+    hpmicro-gpio-port:
```
remove this, `hpmicro-gpio-port` seems to only ever be 1.

willChuai replied:
The io port of the hpm6750evkmini development board is (1)gpiob. As a driver, all possible io configurations should be reserved.

galak commented on drivers/clock_control/Kconfig.hpmicro:
```diff
@@ -0,0 +1,544 @@
+	  Enable driver for Hpmicro Clock Unit.
+
+
+config CLOCK_HPM6750_CLKNAME_LIST
```
where are all these Kconfig symbols used?

willChuai replied:
will delete.

galak reviewed (changes requested):
several DTS related changes needed.

fkokosinski commented:
> > Thanks for these additions. In the meantime the PR has grown quite significantly, could you please keep the change-set in a reasonable size to ease the review process. I'd suggest to limit the first PR to basic SoC support including serial, gpio, and pinctrl drivers, and add further peripherals in follow-up PRs.
> 
> Thanks, great suggestion, I will split the code.

Hi, will the split be still happening, or have you split it already? To be honest, because this PR (and comments) has gotten so large, it's difficult to navigate it.

willChuai commented:
> > > Thanks for these additions. In the meantime the PR has grown quite significantly, could you please keep the change-set in a reasonable size to ease the review process. I'd suggest to limit the first PR to basic SoC support including serial, gpio, and pinctrl drivers, and add further peripherals in follow-up PRs.
> > 
> > 
> > Thanks, great suggestion, I will split the code.
> 
> Hi, will the split be still happening, or have you split it already? To be honest, because this PR (and comments) has gotten so large, it's difficult to navigate it.

Split has been completed，Now only the necessary modules such as gpio, uart, pinctrl, clock are included. Removed pwm, adc, can, spi, i2c modules and another board.

fkokosinski commented on boards/riscv/hpm6750evkmini/Kconfig.defconfig:
```diff
@@ -0,0 +1,17 @@
+config FLASH_LOAD_SIZE
+	default $(dt_chosen_reg_size_hex,$(DT_CHOSEN_Z_CODE_PARTITION))
+
+endif #BOARD_HPM6750EVKMINI
```
Niptick: add space after `#`

fkokosinski commented on boards/riscv/hpm6750evkmini/hpm6750evkmini-pinctrl.dtsi:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2022 hpmicro
```
Please make copyright notices consistent across this PR, and format them in a way that is used across Zephyr, e.g. here it should be `Copyright (c) 2022 HPMicro`.

fkokosinski commented on boards/riscv/hpm6750evkmini/hpm6750evkmini.yaml:
```diff
@@ -0,0 +1,17 @@
+toolchain:
+  - zephyr
+  - cross-compile
+ram: 1024
```
Inconsistent with board docs?

fkokosinski commented on boards/riscv/hpm6750evkmini/support/probes/ft2232.cfg:
```diff
@@ -0,0 +1,15 @@
+ftdi_layout_init 0x0208 0x020b
+ftdi_layout_signal nTRST -data 0x0200 -noe 0x0400
+ftdi_layout_signal nSRST -data 0x0100 -noe 0x0800
+
```
This file (and other cfgs as well) seems to be generic. Is it not provided by openocd by default?

willChuai replied:
These configurations for io are inconsistent from chip to chip.

fkokosinski commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -61,6 +61,8 @@ zephyr_library_sources_ifdef(CONFIG_GPIO_XLNX_PS    gpio_xlnx_ps.c gpio_xlnx_ps_
 zephyr_library_sources_ifdef(CONFIG_GPIO_MCHP_MSS       gpio_mchp_mss.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_SMARTBOND  gpio_smartbond.c)
+zephyr_library_sources_ifdef(CONFIG_GPIO_HPMICRO    gpio_hpmicro.c)
+
```
Remove empty space

fkokosinski commented on drivers/gpio/Kconfig:
```diff
@@ -149,6 +149,7 @@ source "drivers/gpio/Kconfig.sn74hc595"
 source "drivers/gpio/Kconfig.mchp_mss"
 
 source "drivers/gpio/Kconfig.xmc4xxx"
+source "drivers/gpio/Kconfig.hpmicro"
```
Add an empty line to be consistent with how this file is formatted right now

fkokosinski commented on drivers/serial/Kconfig:
```diff
@@ -201,5 +201,6 @@ source "drivers/serial/Kconfig.neorv32"
 source "drivers/serial/Kconfig.xen"
 
 source "drivers/serial/Kconfig.smartbond"
+source "drivers/serial/Kconfig.hpmicro"
```
Ditto

fkokosinski commented on include/zephyr/drivers/clock_control/hpmicro_clock_control.h:
```diff
@@ -0,0 +1,43 @@
+ * @brief Generate clock setting preset
+ *
+ */
+#define HPM_MAKE_SYSCTL_PRESENT(x)		(1 << x)
```
This seems to be identical to the `BIT` macro?

willChuai replied:
The meaning is different, the form is the same. I replaced BIT with.

cfriedt replied:
technically speaking, it should be `(1 << (x))`, but I think the point is why not just 
```
#define HPM_MAKE_SYSCTL_PRESENT(x)		BIT(x)
```

fkokosinski commented on include/zephyr/dt-bindings/pinctrl/hpmicro-pinctrl-common.h:
```diff
@@ -0,0 +1,99 @@
+#define HPMICRO_PIN_ALT_MASK		0x1FU
+#define HPMICRO_PIN_ALT_SHIFT		11U
+#define HPMICRO_PIN_IOC_MASK		0x03U
+#define HPMICRO_PIN_IOC_SHIFT		30U
```
Masks can be defined with `BIT_MASK` macro and other helper macros. Please use them where appropriate.

willChuai commented:
Hi all, please review as soon as possible, split's driver we hope to add in as soon as the first version is approved, either it will make a big difference. Thanks.

cfriedt commented:
@willChuai - just because gpio bugs can be annoying to find later on, can you verify that you have run and passed (at least) the `tests/drivers/gpio/gpio_basic_api` and `tests/drivers/gpio/gpio_api_1pin` suites?

willChuai commented:
> @willChuai - just because gpio bugs can be annoying to find later on, can you verify that you have run and passed (at least) the `tests/drivers/gpio/gpio_basic_api` and `tests/drivers/gpio/gpio_api_1pin` suites?

Only `tests/drivers/gpio/gpio_basic_api` was verified, I will add the relevant code. Another test case verified and replied.

fkokosinski commented on boards/riscv/hpm6750evkmini/support/soc/hpm6750-dual-core.cfg:
```diff
@@ -0,0 +1,60 @@
+# Copyright (c) 2022 HPMicro
```
Where is this file used? Because it doesn't seem to be passed in `board.cmake` to OpenOCD?

willChuai replied:
Set aside for dual-core scenarios. Plan to adapt to dual-core, appear here a little early, delete.

fkokosinski replied:
Doesn't seem to be deleted?

fkokosinski commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,284 @@
+		case GPIO_INT_TRIG_HIGH:
+			trigger = gpio_interrupt_trigger_level_high;
+			break;
+		case GPIO_INT_TRIG_BOTH:
```
We're handing `GPIO_INT_TRIG_BOTH` the same way we handle `default`. Can we merge them?

fkokosinski commented on soc/riscv/riscv-privilege/hpmicro/soc.c:
```diff
@@ -0,0 +1,86 @@
+#endif
+
+#ifdef CONFIG_XIP
+__attribute__ ((section(".nor_cfg_option"))) const uint32_t option[4] = {
```
A comment explaining what this does and why it is needed would be nice.

willChuai replied:
The rom boot of the chip will get the content of this field to initialize the spi flash in the rom.

fkokosinski replied:
Ditto - explanation in this file, not in this convo

fkokosinski commented on soc/riscv/riscv-privilege/hpmicro/start.S:
```diff
@@ -0,0 +1,42 @@
+	call _init_noncache
+
+	li t0, 0xf4002800
+	li t1, 0x3402
```
Explain in comments what these values are.

fkokosinski replied:
Is Wi-Fi enabled by default in this SoC, or is it enabled by something else? Also, would be nice if these values were defined somehow instead of being plain literals.

fkokosinski replied:
BTW I meant that I want this explanation to be in this file as a comment, not in this thread :)

willChuai replied:
Removed this code, called the `hpm_sdk` interface in the clock driver for configuration, and added comments.

willChuai commented:
> > @willChuai - just because gpio bugs can be annoying to find later on, can you verify that you have run and passed (at least) the `tests/drivers/gpio/gpio_basic_api` and `tests/drivers/gpio/gpio_api_1pin` suites?
> 
> Only `tests/drivers/gpio/gpio_basic_api` was verified, I will add the relevant code. Another test case verified and replied.

The above tests have passed. @cfriedt

cfriedt commented:
Looks OK to me from the GPIO side. Please just fix up compliance issues @willChuai

willChuai commented:
Hi all. Is there any plan to complete this pull request so far.

fkokosinski commented:
@willChuai please take a look at the rest of the failing CI checks.

willChuai commented:
> @willChuai please take a look at the rest of the failing CI checks.

Already completed.

cfriedt commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,280 @@
+		return ret;
+	}
+#endif
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
```
Could remove these brackets to simplify `(..) && (..)`

cfriedt commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,280 @@
+		return -ENOTSUP;
+	}
+
+	if (((flags & GPIO_PULL_UP) != 0) || ((flags & GPIO_PULL_DOWN) != 0)) {
```
Could remove these brackets to simplify `(..) || (..)`

cfriedt reviewed (commented):
Some excessive bracket use but otherwise gpio looks ok

fkokosinski commented on soc/riscv/riscv-privilege/hpmicro/Kconfig.soc:
```diff
@@ -0,0 +1,60 @@
+
+endchoice
+
+config SOC_ANDES_V5_HWDSP
```
Agreed - this is a custom extension, so I suppose it should be put in the `arch/riscv` directory, e.g. `arch/riscv/Kconfig.extension` would be a much better place to put this Kconfig option.

willChuai replied:
I removed this feature first in this pull request because it is not currently being used. This also allows for a faster merge into the master.

fkokosinski commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,322 @@
+	int errors = 0;
+
+	return errors;
+}
```
Add `ARG_UNUSED(dev)` and change the return statement to just `return 0;`. A comment explaining that HPM UART doesn't support error checking would also be nice

fkokosinski commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,122 @@
+          - rx00 4 ma
+          - rx01 8 ma
+          - rx11 12 ma
+        required: false
```
Properties are not required by default. This line is redundant.

fkokosinski commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,122 @@
+          - "r110"
+          - "r111"
+      power-source:
+        required: false
```
Ditto

willChuai commented:
Hi all, are there any other questions next?

gmarull commented on modules/hal_hpmicro/Kconfig:
```diff
@@ -0,0 +1,86 @@
+if HAS_HPMICRO_HAL
+
+config USE_HPMSDK_UART
+	bool "HPMSDK UART driver support"
```
these options should be selected by drivers, not directly. so prompt should be removed.

willChuai replied:
Currently it is the selection by driver. For example. `GD` `Infineon` are written in this way.

gmarull commented on modules/hal_hpmicro/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_HAS_HPMICRO_HAL)
+
```
nit: redundant blank line

gmarull commented on west.yml:
```diff
@@ -72,6 +72,12 @@ manifest:
       groups:
         - hal
+    - name: hal_hpmicro
+      url: https://github.com/hpmicro/zephyr_hal_hpmicro.git
```
this HAL comes with code not needed in Zephyr. Why do we need to pull that content? HAL modules are already a mess, let's not make it worse.

ref https://github.com/hpmicro/zephyr_hal_hpmicro/tree/manifest-rev/hpm_sdk/boards/hpm6360evk

willChuai replied:
The entire `board` folder has been deleted

gmarull commented on soc/riscv/riscv-privilege/hpmicro/linker.ld:
```diff
@@ -0,0 +1,35 @@
+#include <zephyr/linker/linker-tool.h>
+#include <zephyr/devicetree.h>
+
+#define IS_CHOSEN_SRAM(x) (DT_DEP_ORD(DT_NODELABEL(x)) == DT_DEP_ORD(DT_CHOSEN(zephyr_sram)))
```
please use `DT_SAME_NODE`.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/pinctrl_soc.h:
```diff
@@ -0,0 +1,64 @@
+#define ZEPHYR_SOC_RISCV_RISCV_PRIVILEGE_HPMICRO_PINCTRL_SOC_H_
+
+#include <zephyr/devicetree.h>
+#include <zephyr/types.h>
```
use `<stdint.h>`

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+#include <zephyr/irq.h>
+#include <hpm_common.h>
+#include <hpm_soc.h>
+#include <board.h>
```
soc code including board specific stuff?

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+#include <hpm_common.h>
+#include <hpm_soc.h>
+#include <board.h>
+#include "hpm_pmp_drv.h"
+#include "hpm_pllctl_drv.h"
+#ifdef CONFIG_XIP
+#include "hpm_bootheader.h"
+#endif
```
these are not local includes, please use `<>`

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+	};
+uint32_t __fw_size__[] = {32768};
+#endif
+__attribute__((weak)) void _init_noncache(void)
```
please use `__weak`

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+	}
+
+	/* Ensure the address and the length are power of 2 aligned */
+	assert((length & (length - 1U)) == 0U);
+	assert((start_addr & (length - 1U)) == 0U);
```
please use `__ASSERT*` macros

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+ *      0 - 4MB / 1 - 8MB / 2 - 16MB
+ */
+__attribute__ ((section(".nor_cfg_option"))) const uint32_t option[4] = {
+		0xfcf90001, 0x00000007, 0x0, 0x0
```
shouldn't all this be in Devicetree?

willChuai replied:
This is a function implemented by MCU ROM, and normally this parameter will not change. And I'm not sure which part to put in the Devicetree is more appropriate.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.h:
```diff
@@ -0,0 +1,18 @@
+#include <soc_common.h>
+#include <zephyr/devicetree.h>
+
+/* Machine timer memory-mapped registers */
+#define RISCV_MTIME_BASE             DT_REG_ADDR_BY_IDX(DT_NODELABEL(mtimer), 0)
+#define RISCV_MTIMECMP_BASE          DT_REG_ADDR_BY_IDX(DT_NODELABEL(mtimer), 1)
```
please use Devicetree for that https://github.com/zephyrproject-rtos/zephyr/pull/48429

gmarull commented on dts/riscv/hpmicro/hpm6750.dtsi:
```diff
@@ -0,0 +1,178 @@
+			interrupts-extended = <&CPU0_intc 3 &CPU1_intc 3>;
+		};
+
+		gpio0: peripheral@f0000000 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			ranges = <0x0 0xf0000000 0x4000>;
+		};
+
+		gpio1: peripheral@f0004000 {
+			#address-cells = <1>;
+			#size-cells = <1>;
+			ranges = <0x0 0xf0004000 0x4000>;
+		};
```
compatible? missing `gpio-controller`? node name doesn't follow dt recommended names either.

willChuai replied:
From the chip point of view, `gpio0` and `gpio1` are the real `gpio controllers`, `gpioa` `gpiob` ... `gpioz` is a grouping of them, if you use `gpio0` `pin100` directly this way to control io, the effect is very poor. The grouping is implemented in `hpm6750evkmini.dts`, `gpiob` is in the `gpio0` controller.

gmarull commented on dts/riscv/hpmicro/hpm6750.dtsi:
```diff
@@ -0,0 +1,178 @@
+			ranges = <0x0 0xf0004000 0x4000>;
+		};
+
+		uart0: serial@f0040000 {
+			compatible = "hpmicro,hpm-uart";
+			reg = <0xf0040000 0x1000>;
+			interrupts = <31 1>;
+			interrupt-parent = <&plic0>;
+			clocks = <&clk HPM_CLOCK_UART0 HPM_CLK_SRC_OSC24M 1>;
+			status = "disabled";
+		};
+
+		pinctrl: pin-controller@f4040000 {
+			compatible = "hpmicro,hpm-pinctrl";
+			reg = <0xf4040000 0x40000>;
+			status = "okay";
+		};
+		pinctrl_pioc: pin-controller@f40d8000 {
+			compatible = "hpmicro,hpm-pinctrl";
+			reg = <0xf40d8000 0x40000>;
+			status = "okay";
+		};
+
+		pinctrl_bioc: pin-controller@f5010000 {
+			compatible = "hpmicro,hpm-pinctrl";
+			reg = <0xf5010000 0x40000>;
+			status = "okay";
+		};
```
please add hw description at the time you add the driver, not before.

willChuai replied:
OK

gmarull commented on dts/bindings/clock/hpmicro,hpm-clock.yaml:
```diff
@@ -0,0 +1,45 @@
+      type: int
+      description: Configure the External OSC ramp-up time. 32KHZ clock
+
+    sysctl-present:
+      required: true
+      type: string
+      description: Clock presets
+      enum:
+        - sysctl-preset-0
+        - sysctl-preset-1
+        - sysctl-preset-2
+        - sysctl-preset-3
```
`present` or `preset`? Also, just use numbers, e.g. `sysctl-preset: 0`.

willChuai replied:
preset

gmarull commented on include/zephyr/drivers/clock_control/hpmicro_clock_control.h:
```diff
@@ -0,0 +1,44 @@
+/** Common clock control device node for all hpmicro chips */
+#define HPMICRO_CLOCK_CONTROL_NODE DT_NODELABEL(clk)
+
+#define CLOCK_HPMICRO_GET_NAME(n) DT_INST_CLOCKS_CELL(n, name)
+#define CLOCK_HPMICRO_GET_SRC(n) DT_INST_CLOCKS_CELL(n, src)
+#define CLOCK_HPMICRO_GET_DIV(n) DT_INST_CLOCKS_CELL(n, div)
```
why wrappers around DT APIs?

gmarull commented on include/zephyr/drivers/clock_control/hpmicro_clock_control.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CLOCK_CONTROL_HPMICRO_CLOCK_CONTROL_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CLOCK_CONTROL_HPMICRO_CLOCK_CONTROL_H_
+
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/dt-bindings/clock/hpmicro-clocks.h>
```
this header is including wrong headers,

```
stdint.h

zephyr/devicetree.h
zephyr/sys/util_macro.h
```

gmarull commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,118 @@
+    };
+    pinmux_gpiob: pinmux_gpiob {
+      group0 {
+        pinmux = <HPMICRO_PINMUX(HPMICRO_PIN(HPMICRO_PORTB, 18),IOC_TYPE_IOC, 0, 0)>,
```
suggestion, since this seems unnecessarily verbose:

```
<HPMICRO_PINMUX(PORTB, 18, IOC, 0, 0)>,
```

(and use macro concatenation within HPMICRO_PINMUX)

gmarull commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,118 @@
+          - rx01 8 ma
+          - rx11 12 ma
+        type: string
+        default: "r000"
```
defaults need to be justified, per bindings policy

gmarull commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,118 @@
+          - "r111"
+      power-source:
+        type: string
+        default: "3v3"
```
ditto

gmarull commented on dts/bindings/pinctrl/hpmicro,hpm-pinctrl.yaml:
```diff
@@ -0,0 +1,118 @@
+          - "r101"
+          - "r110"
+          - "r111"
+      power-source:
```
this is a standard property, see `pincfg-node.yaml`

willChuai replied:
The properties provided by `pincfg-node.yaml` are not very useful, I deal with the same as other manufacturers here such as `NXP`, `Raspberry Pi`, etc.

gmarull commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,322 @@
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+#define PINCTRL_INIT(n) .pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),
+#define PINCTRL_DEFINE(n) PINCTRL_DT_INST_DEFINE(n);
+
+#define HPM_UART_INIT(n)	\
+		PINCTRL_DEFINE(n)	\
```
why this wrapper?

gmarull commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,322 @@
+#define UART_HPMICRO_IRQ_FUNC_DEFINE(n)
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+#define PINCTRL_INIT(n) .pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),
```
same, why?

gmarull commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,280 @@
+	struct gpio_driver_config common;
+	GPIO_Type *gpio_base;
+	uint32_t  port_base;
+#ifdef CONFIG_PINCTRL
+	const struct pinctrl_dev_config *pincfg;
+#endif /* CONFIG_PINCTRL */
```
why is a GPIO driver using pinctrl ?

willChuai replied:
Most of the pins are `gpio` functions by default, but some pins are other functions by default and need to be modified using `pinctrl`, and secondly, drive capability, pull-up and pull-down, etc. also need to be configured by `pinctrl`.

gmarull reviewed (changes requested):
reviewed up to the gpio driver, found several issues, please check my comments.

willChuai committed:
modules: add hpmiro modules

- add hal hpmicro hpm_sdk software packages to zephyr

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai committed:
soc: riscv: add hpmicro series

- add hpmicro series mcu to zephyr

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai committed:
dts: bingings: add hpmciro to vendor-prefixes

- add hpmicro company name to vendor-prefixes

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai committed:
soc: add support for hpmicro hpm6750 soc

- add support for hpmicro hpm6750

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

6750

willChuai committed:
soc: add support for hpmicro hpm6360 soc

- add support for hpmicro hpm6360

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

6360

willChuai committed:
dtsi: add HPMicro hpm6750 device tree

- Adding HPMicro hpm6750 device tree support

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai committed:
drivers: clock_control: add support for hpmicro series mcu

- add hpmicro series mcu clock driver
- add clock dtsi

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai committed:
drivers: pinctrl: add support for hpmicro series mcu

- add pinctrl support for hpmicro
- add pinctrl dtsi

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

pinctrl

willChuai committed:
drivers: serial: add support for hpmicro series mcu

- add hpm-uart driver
- add uart dtsi

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai committed:
drivers: gpio: add support for hpmicro series mcu

- add hpm-gpio driver
- add gpio dtsi

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai committed:
boards: riscv: add support for hpm6750evkmini board

- add support for serial
- add support for gpio
- add usage instructions

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai committed:
tests: drivers: gpio: Add gpio test support for hpm6750evkmini boards

- hpm6750evkmini gpio test passed

Signed-off-by: Yawei CHUAI <yawei.chuai@hpmicro.com>

willChuai commented:
@fkokosinski @gmarull @galak @cfriedt  Modification completed, please continue to review.

cfriedt reviewed (dismissed):
GPIO LGTM

fkokosinski commented:
I see that the CI is failing due to CODEOWNERS check failing. This check will be disabled in #53219, so I don't think that you need to address it in this PR.

fkokosinski commented:
Hi @willChuai, can you rebase your branch to remove conflicts?

willChuai commented:
> Hi @willChuai, can you rebase your branch to remove conflicts?

Hi, @fkokosinski. It's been updated, any specific plans for the back.

fkokosinski commented:
@willChuai thanks for the rebase. Please check the failing _Compliance Checks_ and fix what it has reported.

willChuai commented:
> @willChuai thanks for the rebase. Please check the failing _Compliance Checks_ and fix what it has reported.

Thanks, it has been modified.

willChuai commented:
@galak Does the code satisfy the merge condition?

Cherish-Gww commented:
> @galak Does the code satisfy the merge condition?

sometimes the CI system get an error, maybe it relate with your patch, you can click Update branch or Rebase branch to trigger CI to construct. only all CI build passed, your patch be approved.

fkokosinski commented:
Hi @willChuai, it seems like no changes were introduced in your latest force push. Am I correct here? FYI you only need to rebase if GH reports conflicts. You don't need to do that if e.g. GH reports the out of sync with the main branch message.

willChuai commented:
Please review.

Cherish-Gww commented on soc/riscv/riscv-privilege/hpmicro/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,39 @@
+	default "hpmicro"
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 24000000
```
MCHTIMR clock souce is CLK24MHZ, but the default value is 1MHZ.
I have tested if SYS_CLOCK_HW_CYCLES_PER_SEC is 24MHZ, the k_time 1s is 2.4s.
it need change to the default 1MHZ

willChuai replied:
Thanks for the feedback. The program is designed for 24Mhz, so I will add code to modify the clock division to 1.

Cherish-Gww commented:
@fkokosinski 
Should this patch remove the DNM label？ We've been waiting for a long time

Cherish-Gww commented on soc/riscv/riscv-privilege/hpmicro/start.S:
```diff
@@ -0,0 +1,38 @@
+#endif
+	call _init_noncache
+	/* System reset handler */
+	j __start
```
In RISC-V j instruct only can jump +/- 1M offset,If it exceeds the limit, it will prompt for a link error.
Please change to `call __start`, I have tested call instruct is okay.

fkokosinski commented:
> @fkokosinski Should this patch remove the DNM label？ We've been waiting for a long time

Hi @Cherish-Gww!

The `DNM` label got assigned automatically by zephyrbot because this PR includes changes to external modules. In this case - by adding a new one. The `DNM` label can only be dropped if the new requested HAL module is accepted. If I understand correctly, #46301 needs to go through the TSC and get its approval before that happens, as described in the [Zephyr Project Documentation](https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-a-new-module).

AFAIK, if you want to bring an issue/PR to TSC's attention, you can assign the `TSC` label in a given issue/PR. If I see correctly, the new HAL request issue currently does not have this label assigned to it.

willChuai commented:
> > @fkokosinski Should this patch remove the DNM label？ We've been waiting for a long time
> 
> Hi @Cherish-Gww!
> 
> The `DNM` label got assigned automatically by zephyrbot because this PR includes changes to external modules. In this case - by adding a new one. The `DNM` label can only be dropped if the new requested HAL module is accepted. If I understand correctly, #46301 needs to go through the TSC and get its approval before that happens, as described in the [Zephyr Project Documentation](https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-a-new-module).
> 
> AFAIK, if you want to bring an issue/PR to TSC's attention, you can assign the `TSC` label in a given issue/PR. If I see correctly, the new HAL request issue currently does not have this label assigned to it.

The TSC tag was removed and I can't seem to add it again.

Cherish-Gww commented:
> > > @fkokosinski Should this patch remove the DNM label？ We've been waiting for a long time
> > 
> > 
> > Hi @Cherish-Gww!
> > The `DNM` label got assigned automatically by zephyrbot because this PR includes changes to external modules. In this case - by adding a new one. The `DNM` label can only be dropped if the new requested HAL module is accepted. If I understand correctly, #46301 needs to go through the TSC and get its approval before that happens, as described in the [Zephyr Project Documentation](https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-a-new-module).
> > AFAIK, if you want to bring an issue/PR to TSC's attention, you can assign the `TSC` label in a given issue/PR. If I see correctly, the new HAL request issue currently does not have this label assigned to it.
> 
> The TSC tag was removed and I can't seem to add it again.

I saw in https://github.com/zephyrproject-rtos/zephyr/issues/46301

stephanosio commented on west.yml:
```diff
@@ -79,6 +79,12 @@ manifest:
       path: modules/hal/gigadevice
       groups:
         - hal
+    - name: hal_hpmicro
+      url: https://github.com/hpmicro/zephyr_hal_hpmicro.git
+      revision: 83806f293cf2709e84f4649423c9729b2f507766
+      path: modules/hal/hpmicro
+      groups:
+        - hal
```
https://github.com/zephyrproject-rtos/hal_hpmicro has been created as per https://github.com/zephyrproject-rtos/zephyr/issues/46301.

Please update the manifest to reference the above repository.

willChuai replied:
> https://github.com/zephyrproject-rtos/hal_hpmicro has been created as per #46301.
> 
> Please update the manifest to reference the above repository.

OK, it has been updated. Please start the next step as soon as possible.

willChuai commented:
Hi @fkokosinski , please remove the `DNM` tag.  Please merge as soon as possible too.

fkokosinski commented:
> Hi @fkokosinski , please remove the `DNM` tag. Please merge as soon as possible too.

FYI I don't have merge rights in Zephyr :) Only members of the release team from the Zephyr organization on GitHub have them.

gmarull commented on west.yml:
```diff
@@ -79,6 +79,11 @@ manifest:
       groups:
         - hal
+    - name: hal_hpmicro
+      revision: 83806f293cf2709e84f4649423c9729b2f507766
```
this module seems to include unnecessary stuff:

https://github.com/zephyrproject-rtos/hal_hpmicro/blob/zephyr/hpm_sdk/arch/riscv/riscv_core.h#L17, see include/zephyr/arch/riscv/csr.h
https://github.com/zephyrproject-rtos/hal_hpmicro/blob/zephyr/hpm_sdk/cmake/cmake-ext.cmake, why is this needed?
https://github.com/zephyrproject-rtos/hal_hpmicro/blob/zephyr/hpm_sdk/soc/HPM6750/CMakeLists.txt

Please, make sure to only include stuff used in Zephyr context.

willChuai replied:
`include/zephyr/arch/riscv/csr.h` will be called when the interrupt function is processed.

https://github.com/zephyrproject-rtos/hal_hpmicro/blob/zephyr/hpm_sdk/cmake/cmake-ext.cmake because https://github.com/hpmicro/hpm_sdk 's cmake uses these definitions, hpm_sdk is updated once a quarter, so that we can quickly apply updates to zephyr_hal when hpm_sdk is updated. This approach should not pose any problems.

https://github.com/zephyrproject-rtos/hal_hpmicro/blob/zephyr/hpm_sdk/soc/HPM6750/CMakeLists.txt  This issue is confirmed and will be modified.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/Kconfig.series:
```diff
@@ -0,0 +1,10 @@
+	bool "HPMicro HPM6700_6400 Series"
+	select RISCV
+	select SOC_FAMILY_RISCV_PRIVILEGE
+	select HAS_HPMICRO_HAL
```
A SoC doesn't have a HAL, it's drivers that instead select if they need it.

willChuai replied:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/98945035/5fffa49d-e00e-40d8-9b49-7b9a719feffe)

The soc files will be dependent on the hal, and it looks like other chips handle it that way too.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/pinctrl_soc.h:
```diff
@@ -0,0 +1,64 @@
+ *
+ */
+
+#ifndef ZEPHYR_SOC_RISCV_RISCV_PRIVILEGE_HPMICRO_PINCTRL_SOC_H_
+#define ZEPHYR_SOC_RISCV_RISCV_PRIVILEGE_HPMICRO_PINCTRL_SOC_H_
+
+#include <zephyr/devicetree.h>
+#include <stdint.h>
+#include <zephyr/drivers/pinctrl/pinctrl_hpmicro_common.h>
```
please, add this to the commit adding pinctrl driver.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/sections.ld:
```diff
@@ -0,0 +1,54 @@
+SECTIONS
+{
+#ifdef CONFIG_XIP
+	.nor_cfg_option __nor_cfg_option_load_addr__ : {
+		KEEP(*(.nor_cfg_option))
+	} GROUP_LINK_IN(MCU_BOOT)
+
+	.boot_header __boot_header_load_addr__ : {
+		__boot_header_start__ = .;
+		KEEP(*(.boot_header))
+		KEEP(*(.fw_info_table))
+		KEEP(*(.dc_info))
+		__boot_header_end__ = .;
+	} GROUP_LINK_IN(MCU_BOOT)
+#endif
```
mcuboot stuff here? @nordicjm

Cherish-Gww replied:
> mcuboot stuff here? @nordicjm

XPI NOR boot flash configuration option
It reads this field during the XPI NOR boot process to decide whether to start the firmware

There are comments on this part of the content in the following files
https://github.com/zephyrproject-rtos/zephyr/pull/46329/files#diff-fbd6285e9b4957ff9b98ae797c036a3b42d15ac84ff00b34973fc4d6a78f85bb

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+	};
+uint32_t __fw_size__[] = {32768};
+#endif
+__weak void _init_noncache(void)
```
why weak?

willChuai replied:
The space for noncache is recommended here, and can be configured by the user.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/start.S:
```diff
@@ -0,0 +1,38 @@
+ */
+
+#include <zephyr/toolchain.h>
+#include <soc.h>
```
soc.h? why?

gmarull commented on soc/riscv/riscv-privilege/hpmicro/soc.h:
```diff
@@ -0,0 +1,14 @@
+#define __RISCV_HPMICRO_SOC_H_
+
+#include <soc_common.h>
+#include <zephyr/devicetree.h>
```
please remove DT header from this file.

gmarull commented on soc/riscv/riscv-privilege/hpmicro/vectors.S:
```diff
@@ -0,0 +1,53 @@
+ *
+ */
+
+#include <offsets.h>
+#include <zephyr/toolchain.h>
+
+#include <soc.h>
```
all needed?

willChuai replied:
Y

gmarull commented on dts/riscv/hpmicro/hpm6750.dtsi:
```diff
@@ -0,0 +1,121 @@
+			interrupts-extended = <&CPU0_intc 3 &CPU1_intc 3>;
+	};
+
+	};
```
?

gmarull commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,319 @@
+
+	if (clock_control_configure(clk,
+		(clock_control_subsys_t *) &cfg->uart_clk, NULL) != 0) {
+		return -EIO;
```
please propagate error codes

gmarull commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,319 @@
+	.irq_config_func = irq_config_func##n,
+#define UART_HPMICRO_IRQ_FUNC_DECLARE(n) \
+	static void irq_config_func##n(const struct device *dev);
+#define UART_HPMICRO_IRQ_FUNC_DEFINE(n)	\
+	static void irq_config_func##n(const struct device *dev)	\
+	{	\
+		ARG_UNUSED(dev);	\
+		IRQ_CONNECT(DT_INST_IRQN(n), DT_INST_IRQ(n, priority),	\
+		uart_hpm_isr, DEVICE_DT_INST_GET(n),	\
+		UART_HPMICRO_IRQ_FLAGS(n));		\
+		irq_enable(DT_INST_IRQN(n));	\
+	}
+#else /* !CONFIG_UART_INTERRUPT_DRIVEN */
+#define DEV_CONFIG_IRQ_FUNC_INIT(n)
+#define UART_HPMICRO_IRQ_FUNC_DECLARE(n)
+#define UART_HPMICRO_IRQ_FUNC_DEFINE(n)
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+#define HPM_UART_INIT(n)	\
+		PINCTRL_DT_INST_DEFINE(n);	\
+		UART_HPMICRO_IRQ_FUNC_DECLARE(n)	\
+		static struct uart_hpm_data uart_hpm_data_##n = {	\
+		.baud_rate = DT_INST_PROP(n, current_speed),	\
+	};									\
+	static const struct uart_hpm_cfg uart_hpm_config_##n = {	\
+		.base = (UART_Type *)DT_INST_REG_ADDR(n),	\
+		.parity = DT_INST_ENUM_IDX_OR(n, parity, UART_CFG_PARITY_NONE),	\
+		DEV_CONFIG_IRQ_FUNC_INIT(n)		\
+		.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),		\
+		HPM_CLOCK_CFG_DATA_DEFAULT(uart_clk, n)	\
+	};	\
+	DEVICE_DT_INST_DEFINE(n, &uart_hpm_init,	\
+				  NULL,	\
+				  &uart_hpm_data_##n,	\
+				  &uart_hpm_config_##n, PRE_KERNEL_1,	\
+				  CONFIG_SERIAL_INIT_PRIORITY,	\
+				  &uart_hpm_driver_api);	\
+	UART_HPMICRO_IRQ_FUNC_DEFINE(n)
+
+DT_INST_FOREACH_STATUS_OKAY(HPM_UART_INIT)
```
please improve formatting here, this is unreadable

gmarull commented on drivers/serial/uart_hpmicro.c:
```diff
@@ -0,0 +1,319 @@
+#define UART_HPMICRO_IRQ_FUNC_DECLARE(n) \
+	static void irq_config_func##n(const struct device *dev);
+#define UART_HPMICRO_IRQ_FUNC_DEFINE(n)	\
+	static void irq_config_func##n(const struct device *dev)	\
+	{	\
+		ARG_UNUSED(dev);	\
```
why is this taking a device that isn't used?

gmarull commented on drivers/gpio/gpio_hpmicro.c:
```diff
@@ -0,0 +1,280 @@
+	GPIO_Type *gpio_base = config->gpio_base;
+	uint32_t port_base = config->port_base;
+
+#ifdef CONFIG_PINCTRL
+	int ret;
+
+	ret = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+#endif
```
why is a GPIO driver using pinctrl? Can you explain?

nordicjm commented on soc/riscv/riscv-privilege/hpmicro/sections.ld:
```diff
@@ -0,0 +1,54 @@
+#include <zephyr/linker/linker-tool.h>
+#include <zephyr/devicetree.h>
+
+#ifdef CONFIG_XIP
+__nor_cfg_option_load_addr__ = DT_REG_ADDR(DT_CHOSEN(zephyr_flash)) + 0x400;
+__boot_header_load_addr__ = DT_REG_ADDR(DT_CHOSEN(zephyr_flash)) + 0x1000;
+__app_load_addr__ = DT_REG_ADDR(DT_CHOSEN(zephyr_flash)) + 0x3000;
+__boot_header_length__ = __boot_header_end__ - __boot_header_start__;
+__app_offset__ = __app_load_addr__ - __boot_header_load_addr__;
+#endif
+
+SECTIONS
+{
+#ifdef CONFIG_XIP
+	.nor_cfg_option __nor_cfg_option_load_addr__ : {
+		KEEP(*(.nor_cfg_option))
+	} GROUP_LINK_IN(MCU_BOOT)
+
+	.boot_header __boot_header_load_addr__ : {
+		__boot_header_start__ = .;
+		KEEP(*(.boot_header))
+		KEEP(*(.fw_info_table))
+		KEEP(*(.dc_info))
+		__boot_header_end__ = .;
+	} GROUP_LINK_IN(MCU_BOOT)
```
What is the reasoning for any of this?

willChuai replied:
MCU BOOT here refers to the chip's own BOOT, not the MCUBoot software, which I renamed here.

willChuai commented:
Please remove the DNM tag, why this tag is automatically added, and whether the auto-add feature can be removed.

fkokosinski commented:
> Please remove the DNM tag, why this tag is automatically added, and whether the auto-add feature can be removed.

It looks like the DNM label is added automatically because the `hal_hpmicro` is currently not present in the `west.yml` file on the main branch, and this triggers Zephyrbot to add the DNM label: https://github.com/zephyrproject-rtos/zephyr/actions/runs/5121216514/jobs/9208697947?pr=46329#step:3:104.

It seems that this is handled by the [manifest](https://github.com/zephyrproject-rtos/action-manifest) action, but I'm not sure if its current behavior is intentional. This means that every time you do a push, and by doing this - trigger the CI, the DNM label will be reassigned.

gmarull reviewed (changes requested):
HAL continues to include unnecessary stuff, e.g. https://github.com/zephyrproject-rtos/hal_hpmicro/tree/zephyr/hpm_sdk/arch please clean it up

willChuai commented:
> HAL continues to include unnecessary stuff, e.g. https://github.com/zephyrproject-rtos/hal_hpmicro/tree/zephyr/hpm_sdk/arch please clean it up

Thanks, modifying, pushing soon.

willChuai commented:
> HAL continues to include unnecessary stuff, e.g. https://github.com/zephyrproject-rtos/hal_hpmicro/tree/zephyr/hpm_sdk/arch please clean it up

Thanks, modifying, pushing soon.

willChuai commented:
> HAL continues to include unnecessary stuff, e.g. https://github.com/zephyrproject-rtos/hal_hpmicro/tree/zephyr/hpm_sdk/arch please clean it up

It's done.

fkokosinski commented:
@willChuai please resolve conflicts

fkokosinski commented on west.yml:
```diff
@@ -159,6 +159,11 @@ manifest:
       groups:
         - hal
+    - name: hal_hpmicro
+      revision: pull/2/head
```
Re-added the DNM label because this now points to a PR: https://github.com/zephyrproject-rtos/hal_hpmicro/pull/2

Cherish-Gww commented:
1 year of waiting

cfriedt reviewed (dismissed):
LGTM for GPIO

Cherish-Gww commented:
@willChuai 
have you tested CONFIG_XIP?
I'm a bit confused about flash partitioning, He differs from the description in the user manual.

Cherish-Gww commented on soc/riscv/riscv-privileged/hpmicro/sections.ld:
```diff
@@ -0,0 +1,54 @@
+		KEEP(*(.start))
+	} GROUP_LINK_IN(ROMABLE_REGION)
+
+	SECTION_PROLOGUE(.noncacheable,,)
```
custom RAM section no need to load, SECTION_PROLOGUE(.noncacheable,(NOLOAD),)

Cherish-Gww replied:
Please see: https://github.com/zephyrproject-rtos/zephyr/pull/46329#discussion_r1278845561

Cherish-Gww commented on soc/riscv/riscv-privileged/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+#ifdef CONFIG_XIP
+/**
+ * @brief FLASH configuration option definitions:
+ * option[0]:
```
According to the description in the user manual:
this word should be Header, not option[0]

Cherish-Gww commented on soc/riscv/riscv-privileged/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+ *    [31:16] 0xfcf9 - FLASH configuration option tag
+ *    [15:4]  0 - Reserved
+ *    [3:0]   option words (exclude option[0])
+ * option[1]:
```
According to the description in the user manual:
this word should be option[0], not option[1]

Cherish-Gww commented on soc/riscv/riscv-privileged/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+ *      1 - 30MHz / 2 - 50MHz / 3 - 66MHz / 4 - 80MHz / 5 - 100MHz
+ *				/ 6 - 120MHz / 7 - 133MHz / 8 - 166MHz
+ *
+ * option[2] (Effective only if the bit[3:0] in option[0] > 1)
```
According to the description in the user manual:
this should be option[1], not option[2]

Cherish-Gww commented on soc/riscv/riscv-privileged/hpmicro/soc.c:
```diff
@@ -0,0 +1,143 @@
+ *			(Two FLASH connected to CA and CB respectively)
+ *    [7:0] Drive Strength
+ *      0 - Default value
+ * option[3] (Effective only if the bit[3:0] in option[0] > 2,
```
According to the description in the user manual:
it should be option[2], not option[3]

Cherish-Gww commented on dts/riscv/hpmicro/hpm6750.dtsi:
```diff
@@ -0,0 +1,178 @@
+			reg = <0x1180000 DT_SIZE_K(256)>;
+		};
+
+		noncache: memory@11c0000 {
```
I don't think it's necessary to establish a noncache node.
Using zephyr/arch/common/nocache.ld instead, it will automatic distribution on demand. I have tested by using nocache.ld in ethernet driver.

Cherish-Gww commented on soc/riscv/riscv-privileged/hpmicro/Kconfig.soc:
```diff
@@ -0,0 +1,55 @@
+	select ATOMIC_OPERATIONS_BUILTIN
+	select INCLUDE_RESET_VECTOR
+	select RISCV_ISA_EXT_M
+	select RISCV_ISA_EXT_A
```
According to the user manual description HPM6750 support:
RV32I, M, A, F, D, C, P(SIMD and DSP)
so you have lost:
RISCV_ISA_EXT_F
RISCV_ISA_EXT_D
RISCV_ISA_EXT_C

ddavidebor commented:
Hey HPMicro! Just wanted to let you know we're excited about this PR at my company, and we look forward to buy lots and lots of your microcontrollers !

(without Zephyr support, we wouldn't buy them)

Cherish-Gww commented:
I organized this PR here: https://github.com/zephyrproject-rtos/zephyr/pull/64117
I will follow this PR every day

Cherish-Gww commented:
> Hey HPMicro! Just wanted to let you know we're excited about this PR at my company, and we look forward to buy lots and lots of your microcontrollers !
> 
> (if no Zephyr support, we will use micros from other companies)

I am using hpmicro chip too,  I have debugged the vast majority of peripheral drivers.

Cherish-Gww commented on drivers/pinctrl/CMakeLists.txt:
```diff
@@ -34,4 +34,8 @@ zephyr_library_sources_ifdef(CONFIG_PINCTRL_TI_K3 pinctrl_ti_k3.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_EMSDP pinctrl_emsdp.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_TI_CC32XX pinctrl_ti_cc32xx.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_NUMAKER pinctrl_numaker.c)
+<<<<<<< HEAD
```
Conflict introduced

Cherish-Gww commented on drivers/pinctrl/CMakeLists.txt:
```diff
@@ -34,4 +34,8 @@ zephyr_library_sources_ifdef(CONFIG_PINCTRL_TI_K3 pinctrl_ti_k3.c)
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_NUMAKER pinctrl_numaker.c)
+<<<<<<< HEAD
 zephyr_library_sources_ifdef(CONFIG_PINCTRL_QUICKLOGIC_EOS_S3 pinctrl_eos_s3.c)
+=======
```
Please keep them all

Cherish-Gww commented on drivers/pinctrl/Kconfig:
```diff
@@ -63,6 +63,10 @@ source "drivers/pinctrl/Kconfig.ti_k3"
 source "drivers/pinctrl/Kconfig.numaker"
+<<<<<<< HEAD
 source "drivers/pinctrl/Kconfig.eos_s3"
+=======
```
Please keep them all

willChuai commented:
I will create a new pull request because the current pull request process is taking too long and is too confusing.
